File: examples/luajit-http/transport.lua
local ffi = require "ffi"
local json = require "cjson.safe"

ffi.cdef [[
typedef struct StogasTransport StogasTransport;
uint32_t stogas_verifier_abi_version(void);
char *stogas_transport_start(const char *, size_t, StogasTransport **);
void stogas_transport_close(const StogasTransport *);
void stogas_transport_free(StogasTransport *);
void stogas_verifier_string_free(char *);
]]

return function(library, configuration)
    local native = ffi.load(library)
    assert(native.stogas_verifier_abi_version() == 1, "Unsupported native ABI")
    local config = configuration and assert(json.encode(configuration)) or "{}"
    local output = ffi.new("StogasTransport *[1]")
    local raw = native.stogas_transport_start(config, #config, output)
    local result = raw ~= nil and json.decode(ffi.string(raw)) or nil
    native.stogas_verifier_string_free(raw)
    if not result or result.ok ~= true or output[0] == nil or
       type(result.value) ~= "table" or type(result.value.base_url) ~= "string" then
        native.stogas_transport_free(output[0])
        error("Unable to start verified transport")
    end
    local handle = ffi.gc(output[0], function(value) native.stogas_transport_free(value) end)
    return {
        base_url = result.value.base_url,
        close = function()
            if handle ~= nil then
                ffi.gc(handle, nil)
                native.stogas_transport_close(handle)
                native.stogas_transport_free(handle)
                handle = nil
            end
        end,
    }
end

File: examples/php-openai/Transport.php
<?php
declare(strict_types=1);

final class Transport
{
    private FFI $native;
    private ?FFI\CData $handle = null;
    public readonly string $baseUrl;

    public function __construct(string $library, array $configuration = [])
    {
        $this->native = FFI::cdef(<<<'C'
            typedef struct StogasTransport StogasTransport;
            uint32_t stogas_verifier_abi_version(void);
            char *stogas_transport_start(const char *, size_t, StogasTransport **);
            void stogas_transport_close(const StogasTransport *);
            void stogas_transport_free(StogasTransport *);
            void stogas_verifier_string_free(char *);
            C, $library);
        if ($this->native->stogas_verifier_abi_version() !== 1) {
            throw new RuntimeException('Unsupported Stogas native ABI');
        }
        $this->handle = $this->native->new('StogasTransport *');
        $config = json_encode((object) $configuration, JSON_THROW_ON_ERROR);
        $raw = $this->native->stogas_transport_start($config, strlen($config), FFI::addr($this->handle));
        try {
            if ($raw === null || FFI::isNull($raw)) {
                throw new RuntimeException('Unable to start verified transport');
            }
            $result = json_decode(FFI::string($raw), true, flags: JSON_THROW_ON_ERROR);
            if (($result['ok'] ?? false) !== true || FFI::isNull($this->handle) ||
                !is_string($result['value']['base_url'] ?? null)) {
                throw new RuntimeException('Unable to start verified transport');
            }
            $this->baseUrl = $result['value']['base_url'];
        } catch (Throwable $error) {
            $this->close();
            throw $error;
        } finally {
            $this->native->stogas_verifier_string_free($raw);
        }
    }

    public function close(): void
    {
        if ($this->handle !== null && !FFI::isNull($this->handle)) {
            $this->native->stogas_transport_close($this->handle);
            $this->native->stogas_transport_free($this->handle);
        }
        $this->handle = null;
    }

    public function __destruct() { $this->close(); }
    private function __clone() {}
    public function __serialize(): array { throw new LogicException('Native transport cannot be serialized'); }
}
